osquery/tables/system/windows/windows_crashes.cpp

```diff
@@ -0,0 +1,764 @@
+    {0x00010000, "MiniDumpWithPrivateWriteCopyMemory"},
+    {0x00020000, "MiniDumpIgnoreInaccessibleMemory"},
+    {0x00040000, "MiniDumpWithTokenInformation"},
+    {0x00080000, "MiniDumpWithModuleHeaders"},
+    {0x00100000, "MiniDumpFilterTriage"},
+    {0x001fffff, "MiniDumpValidTypeFlags"}};
+const std::vector<MINIDUMP_STREAM_TYPE> kStreamTypes = {ExceptionStream,
+                                                        ModuleListStream,
+                                                        ThreadListStream,
+                                                        MemoryListStream,
+                                                        SystemInfoStream,
+                                                        MiscInfoStream};
+
+Status logAndStoreTID(const MINIDUMP_EXCEPTION_STREAM* stream,
+                      unsigned int& tid,
+                      Row& r) {
+  if ((stream == nullptr) || (stream->ThreadId == 0))
+    return Status(1);
+
+  r["tid"] = BIGINT(stream->ThreadId);
+  tid = stream->ThreadId;
+  return Status(0);
+}
+
+Status logAndStoreExceptionCodeAndAddr(const MINIDUMP_EXCEPTION_STREAM* stream,
+                                       unsigned int& exCode,
+                                       unsigned long long& exAddr,
+                                       Row& r) {
+  if (stream == nullptr)
+    return Status(1);
+
+  auto ex = stream->ExceptionRecord;
+  std::ostringstream exCodeStr;
+  exCodeStr << "0x" << std::hex << ex.ExceptionCode;
+  if (exCodeStr.fail())
+    return Status(1);
+  r["exception_code"] = exCodeStr.str();
+  exCode = ex.ExceptionCode;
+
+  std::ostringstream exAddrStr;
+  exAddrStr << "0x" << std::hex << ex.ExceptionAddress;
+  if (exAddrStr.fail())
+    return Status(1);
+  r["exception_address"] = exAddrStr.str();
+  exAddr = ex.ExceptionAddress;
+  return Status(0);
+}
+
+/*
+Log the exception message for errors with defined parameters
+(see ExceptionInformation @
+https://msdn.microsoft.com/en-us/library/windows/desktop/ms680367(v=vs.85).aspx)
+*/
+Status logExceptionMsg(const MINIDUMP_EXCEPTION_STREAM* stream,
+                       unsigned int exCode,
+                       unsigned long long exAddr,
+                       Row& r) {
+  if (stream == nullptr)
+    return Status(1);
+
+  auto ex = stream->ExceptionRecord;
+  if ((ex.ExceptionCode == EXCEPTION_ACCESS_VIOLATION) &&
+      (ex.NumberParameters == 2)) {
+    std::ostringstream errorMsg;
+    std::ostringstream memAddrStr;
+    memAddrStr << "0x" << std::hex << ex.ExceptionInformation[1];
+    std::ostringstream exAddrStr;
+    exAddrStr << "0x" << std::hex << exAddr;
+
+    errorMsg << "The instruction at " << exAddrStr.str()
+             << " referenced memory at " << memAddrStr.str() << ".";
+    switch (ex.ExceptionInformation[0]) {
+    case 0:
+      errorMsg << " The memory could not be read.";
+      break;
+    case 1:
+      errorMsg << " The memory could not be written.";
+      break;
+    case 8:
```

Comment:
Checking the MSDN articles about the `ExceptionInformation` stream, I'm only seeing references to `0` and `1` for valid codes. Is there another exhaustive list to read up on all possible code values? Should there be more cases in this switch statement? If that's the case, it might be better to restructure this to be even more terse, and just have `r["exception_message"]` be the integer value indicating the exception code encountered, and leave it to the user to go lookup the message correlating to the value. Thoughts?

Reply:
DEP info is on https://msdn.microsoft.com/en-us/library/windows/desktop/aa363082(v=vs.85).aspx under ExceptionInformation (which I just found right now, this should really be more consistent across MS's documentation...). I see where you're going on just having the error code there though, my approach was to try to automate the research out so we can deliver human-understandable results right off the bat :) Looks like there are only these 3 cases though, so it's not like we have to build an insane switch statement for everything. I'd be game for leaving the message strings in, but your call :)

Reply:
Yeah I'd say the same thing here as I mentioned above. While it does save a little bit of time to the user, I think that most of these exception codes are pretty normally google'd. For example, just taking from the MSDN article you linked:
![screen shot 2017-10-06 at 1 02 33 pm](https://user-images.githubusercontent.com/1328916/31296305-a92ebee4-aa96-11e7-851a-2817a7c5930f.png)
If we grab those values and throw them to a quick google, you'll see values for `0xc0000005` and `0xC0000006 `, which are super common error codes in Windows. I'll leave this decision to you, on how you'd like the data presented, but my two-cents on this is that it's better for us to start with lower amounts of information, such as just the error code, and allow the user to look that data up, and then if we see a need for more data to be added, bring that into the fold with future PRs.